dataRetriever/dataPool/headersCache/headersNonceCache.go

```diff
@@ -0,0 +1,252 @@
+		headerHash: headerHash,
+		header:     header,
+	}
+	headerListD.headerList = append(headerListD.headerList, headerDetails)
+	headerShardPool.addElement(headerNonce, headerListD)
+	hnc.mutHeadersMap.Unlock()
+
+	hnc.hdrsCounter.increment(headerShardId)
+
+	hnc.tryToDoEviction(headerShardId)
+
+	return false
+
+}
+
+func (hnc *headersNonceCache) getShardMap(shardId uint32) *headersMap {
+	hnc.mutHeadersMap.Lock()
+	defer hnc.mutHeadersMap.Unlock()
+
+	if _, ok := hnc.hdrNonceCache[shardId]; !ok {
+		hnc.hdrNonceCache[shardId] = newHeadersMap()
+		hnc.canDoEviction[shardId] = make(chan struct{}, 1)
+	}
+
+	return hnc.hdrNonceCache[shardId]
+}
+
+func (hnc *headersNonceCache) removeHeaderNonceCache(hdrInfo headerInfo, headerHash []byte) {
+	_, ok := hnc.hdrNonceCache[hdrInfo.headerShardId]
+	if !ok {
+		return
+	}
+
+	hdrListD, ok := hnc.hdrNonceCache[hdrInfo.headerShardId].getHeadersDetailsListFromSMap(hdrInfo.headerNonce)
+	if !ok {
+		return
+	}
+
+	//remove header from header list
+	for index, headerD := range hdrListD.headerList {
+		if !bytes.Equal(headerD.headerHash, headerHash) {
+			continue
+		}
+
+		hdrListD.headerList = append(hdrListD.headerList[:index], hdrListD.headerList[index+1:]...)
+		hnc.hdrsCounter.decrement(hdrInfo.headerShardId)
+
+		if len(hdrListD.headerList) == 0 {
+			hnc.hdrNonceCache[hdrInfo.headerShardId].removeElement(hdrInfo.headerNonce)
+			return
+		}
+
+		hnc.hdrNonceCache[hdrInfo.headerShardId].addElement(hdrInfo.headerNonce, hdrListD)
+		return
+	}
+}
+
+func (hnc *headersNonceCache) removeHeaderNonceByNonceAndShardId(hdrNonce uint64, shardId uint32) int {
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return 0
+	}
+
+	hdrListD, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return 0
+	}
+
+	hdrsHashes := make([][]byte, 0)
+	for _, hdrDetails := range hdrListD.headerList {
+		hdrsHashes = append(hdrsHashes, hdrDetails.headerHash)
+		hnc.hdrsCounter.decrement(shardId)
+	}
+
+	hnc.hdrNonceCache[shardId].removeElement(hdrNonce)
+
+	//remove elements from hashes map
+	for _, hash := range hdrsHashes {
+		hnc.headersByHash.deleteElement(hash)
+	}
+
+	return len(hdrsHashes)
+}
+
+func (hnc *headersNonceCache) getHeadersByNonceAndShardId(hdrNonce uint64, shardId uint32) ([]headerDetails, bool) {
+	hnc.mutHeadersMap.RLock()
+	defer hnc.mutHeadersMap.RUnlock()
+
+	_, ok := hnc.hdrNonceCache[shardId]
+	if !ok {
+		return nil, false
+	}
+
+	headersList, ok := hnc.hdrNonceCache[shardId].getHeadersDetailsListFromSMap(hdrNonce)
+	if !ok {
+		return nil, false
+	}
+
+	return headersList.headerList, true
+}
+
+func (hnc *headersNonceCache) getNumHeaderFromCache(shardId uint32) int64 {
+	return hnc.hdrsCounter.getNumHeaderFromCache(shardId)
+}
+
+func (hnc *headersNonceCache) lruEviction(shardId uint32) {
+	nonces := hnc.hdrNonceCache[shardId].getNoncesTimestampSorted()
+
+	var numHashes int
+	for i := 0; i < hnc.numHeadersToRemove && i < len(nonces); i++ {
```

Comment:
Add variable called `maxItemsToRemove` perhaps? Min between `numHeadersToRemove` and `len(nonces)`?

Reply:
done.